apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/ir/BackendIrBuilder.kt

```diff
@@ -1,205 +1,298 @@
-import com.apollographql.apollo.compiler.frontend.ir.Condition
-import com.apollographql.apollo.compiler.frontend.ir.Field
-import com.apollographql.apollo.compiler.frontend.ir.Fragment
-import com.apollographql.apollo.compiler.frontend.ir.FragmentRef
-import com.apollographql.apollo.compiler.frontend.ir.InlineFragment
-import com.apollographql.apollo.compiler.frontend.ir.Operation
-import com.apollographql.apollo.compiler.frontend.ir.SourceLocation
+import com.apollographql.apollo.compiler.frontend.gql.GQLDirective
+import com.apollographql.apollo.compiler.frontend.gql.GQLField
+import com.apollographql.apollo.compiler.frontend.gql.GQLFieldDefinition
+import com.apollographql.apollo.compiler.frontend.gql.GQLFragmentDefinition
+import com.apollographql.apollo.compiler.frontend.gql.GQLFragmentSpread
+import com.apollographql.apollo.compiler.frontend.gql.GQLInlineFragment
+import com.apollographql.apollo.compiler.frontend.gql.GQLNamedType
+import com.apollographql.apollo.compiler.frontend.gql.GQLOperationDefinition
+import com.apollographql.apollo.compiler.frontend.gql.GQLSelectionSet
+import com.apollographql.apollo.compiler.frontend.gql.GQLType
+import com.apollographql.apollo.compiler.frontend.gql.GQLTypeDefinition
+import com.apollographql.apollo.compiler.frontend.gql.GQLVariableValue
+import com.apollographql.apollo.compiler.frontend.gql.Schema
+import com.apollographql.apollo.compiler.frontend.gql.SourceLocation
+import com.apollographql.apollo.compiler.frontend.gql.definitionFromScope
+import com.apollographql.apollo.compiler.frontend.gql.findDeprecationReason
+import com.apollographql.apollo.compiler.frontend.gql.leafType
+import com.apollographql.apollo.compiler.frontend.gql.possibleTypes
+import com.apollographql.apollo.compiler.frontend.gql.responseName
+import com.apollographql.apollo.compiler.frontend.gql.rootTypeDefinition
+import com.apollographql.apollo.compiler.frontend.gql.schemaKind
+import com.apollographql.apollo.compiler.frontend.gql.toKotlinValue
+import com.apollographql.apollo.compiler.frontend.gql.toSchemaType
+import com.apollographql.apollo.compiler.frontend.gql.toUtf8WithIndents
+import com.apollographql.apollo.compiler.frontend.gql.usedFragmentNames
+import com.apollographql.apollo.compiler.frontend.gql.validateAndCoerce
 import com.apollographql.apollo.compiler.introspection.IntrospectionSchema
-import com.apollographql.apollo.compiler.introspection.possibleTypes
-import com.apollographql.apollo.compiler.introspection.resolveType
 
 internal class BackendIrBuilder private constructor(
-    private val schema: IntrospectionSchema,
-    private val namedFragments: List<Fragment>,
+    private val schema: Schema,
+    private val fragmentDefinitions: Map<String, GQLFragmentDefinition>,
     private val useSemanticNaming: Boolean,
+    val packageNameProvider: PackageNameProvider?,
+    val typesPackageName: String,
+    val fragmentsPackageName: String
 ) {
+  data class BackendIrBuilderInput(
```

Comment:
Looks like we can't avoid having some abstraction that will be an input for backend IR builder. What do you think of moving this class out of `BackendIrBuilder` and call it `FrontendIR`. Also we can rename `IRBuilder` to `FrontendIRBuilder`?

Reply:
Yep, we won't save the intermediate abstraction there. I don't think "IR" matches well though because it will still contain the `GQLDocuments`. What this abstraction brings is: 

* `fragmentsToGenerate`, `enumToGenerate`, `inputTypeToGenerate`
* packageName stuff
* maybe `sourceWithFragment` 

So it's more about "augmenting" the GQLDocument than transforming it to an intermediate representation. But I don't really have good alternative suggestions either.

Reply:
Right, maybe IR is not good naming here and we should come up with some better naming, as we won't just pass it to BackendIR but also run some optimization like squasing inline fragments, merging fields etc.

Reply:
See https://github.com/apollographql/apollo-android/pull/2789, I added a `TypesToGenerate` class because this is the main information actually computed by `IRBuilder`. 

squasing inline fragments, merging fields etc, are transformations on the GQLDocument so I don't think we need an intermediate data structure for this. GraphQLCompiler is growing a bit with https://github.com/apollographql/apollo-android/pull/2789 but I don't mind this too much. We need a central place where the compiler applies the different steps.